scrapypuppeteer/middleware.py

```diff
@@ -164,3 +166,120 @@ def close_used_contexts(self, spider):
                               meta={"proxy": None},
                               body=json.dumps(contexts))
             return self.crawler.engine.downloader.fetch(request, None)
+
+
+class PuppeteerRecaptchaDownloaderMiddleware:
+    """
+    This middleware is supposed to solve recaptcha on the page automatically.
+    If there is no captcha on the page then this middleware will do nothing on
+    the page, so your 2captcha balance will remain the same.
+    It can submit recaptcha if "submit button" is provided.
+    It will not "submit" captcha if there is no submit-selector.
+
+    Settings:
+
+    RECAPTCHA_SOLVING: bool = True - whether solve captcha automatically or not
+    SUBMIT_RECAPTCHA_SELECTOR: dict = {} - dictionary consisting of domains and
+        these domains' submit selectors, e.g.
+            'www.google.com/recaptcha/api2/demo': '#recaptcha-demo-submit'
+        it could be also squeezed to
+            'ecaptcha/api2/de': '#recaptcha-demo-submit'
+        In general - unique identifying string which is contained in web-page url
+        If there is no button to submit recaptcha then provide empty string to a domain.
+    """
+
+    RECAPTCHA_SOLVING_SETTING = "RECAPTCHA_SOLVING"
+    SUBMIT_SELECTORS_SETTING = "SUBMIT_RECAPTCHA_SELECTORS"
+
+    def __init__(self,
+                 recaptcha_solving: bool,
+                 submit_selectors: dict):
+        self.submit_selectors = submit_selectors
+        self.recaptcha_solving = recaptcha_solving
+        self._page_responses = dict()
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler):
+        recaptcha_solving = crawler.settings.get(cls.RECAPTCHA_SOLVING_SETTING, True)
+        submit_selectors = crawler.settings.getdict(cls.SUBMIT_SELECTORS_SETTING, dict())
+        if recaptcha_solving and not submit_selectors:
+            raise ValueError('No submit selectors provided but automatic solving is enabled')
+        return cls(recaptcha_solving, submit_selectors)
+
+    @staticmethod
+    def process_request(request, spider):
+        # We don't modify any request, we only work with responses
+        return None
+
+    def process_response(self,
+                         request, response,
+                         spider):
+        # TODO: What if we did not provide 2captcha token?
+        if not isinstance(response, PuppeteerResponse):  # We only work with PuppeteerResponses
+            return response
+
+        if response.puppeteer_request.close_page:  # No need in solving captcha right before closing the page
+            return response
+
+        if isinstance(response.puppeteer_request.action, Click) and \
+                response.puppeteer_request.action.selector in self.submit_selectors.values():  # Submitted captcha
+            return self._gen_response(response)
```

Comment:
In general this action can be a regular click on some unrelated to captchas element. It is ok to expect submit selectors to be unique, But instead you can track this was indeed your captcha sumbission request - for example, you can pass some `'_is_captcha_submission'` flag with `request.meta`.